### PR TITLE
More Info: Add History Tab

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:
- This PR requires backend: home-assistant/core#39555

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 1]
```diff
@@ -286,6 +286,7 @@ export class MoreInfoDialog extends LitElement {
       this._fetchUserDone,
     ]);
 
+    // @ts-ignore
     this._entries = entries.slice(0, 5);
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts]: Should we slice it to 5 or show more and add a overflow?

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 2]
```diff
@@ -93,85 +115,133 @@ export class MoreInfoDialog extends LitElement {
...
-                  )}
-                </p>
-                <mwc-button class="warning" @click=${this._removeEntity}>
-                  ${this.hass.localize(
-                    "ui.dialogs.more_info_control.restored.remove_action"
-                  )}
-                </mwc-button>`
-            : ""}
+            : html`
+                ${this._computeShowHistoryComponent(entityId)
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts:215]: We should not show the tab in this case

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 3]
```diff
@@ -57,7 +74,9 @@ export class MoreInfoDialog extends LitElement {
     this.large = false;
     this._stateHistory = undefined;
     if (this._computeShowHistoryComponent(this._entityId)) {
+      this._fetchUserDone = this._fetchUserNames();
       this._getStateHistory();
+      this._getLogBookData();
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts:79]: We should fetch this data when the history tab is activated. Extract all history and logbook code and bring it to a new component `ha-dialog-tab-history`. We should then lazy load that component after the dialog is opened.

**Comment by bramkragten** [comment]
For the layout, can we put the time above the line for the logbook?

**Comment by zsarnett** [comment]
Yea I can probably add a narrow layout to the log book

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 4]
```diff
@@ -198,6 +268,63 @@ export class MoreInfoDialog extends LitElement {
...
+  private async _getLogBookData() {
+    if (!this._entityId) {
+      return;
+    }
+    this._isLoading = true;
+
+    const yesterday = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
+    const now = new Date();
+    const [entries] = await Promise.all([
+      getLogbookData(
```
Comment (bdraco) [inline src/dialogs/more-info/ha-more-info-dialog.ts:280]: We will need to teach `getLogbookData` to pass the `entity_matches_only ` https://github.com/home-assistant/core/pull/39555

Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-dialog.ts:280]: Sounds good. Thanks @bdraco

**Comment by zsarnett** [comment]
image

File: src/data/logbook.ts [hunk 5]
```diff
@@ -60,11 +62,13 @@ const getLogbookDataFromServer = async (
...
   startDate: string,
   endDate: string,
-  entityId?: string
+  entityId?: string,
+  entity_matches_only?: boolean
 ) => {
   const url = `logbook/${startDate}?end_time=${endDate}${
     entityId ? `&entity=${entityId}` : ""
-  }`;
+  }${entity_matches_only ? `&entity_matches_only=true` : ""}`;
```
Comment (bdraco) [inline src/data/logbook.ts:70]:
```suggestion
  }${entity_matches_only ? `&entity_matches_only` : ""}`;
```

**Comment by bdraco** [comment]
Also no harm in merging this before the backend is done as it will just be ~2x slower until the backend change is done.

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 6]
```diff
@@ -243,6 +244,19 @@ export class MoreInfoDialog extends LitElement {
...
 
+  private _handleTabChanged(ev: CustomEvent): void {
+    const newTab = ev.detail.index;
+    if (newTab === this._currTabIndex) {
+      return;
+    }
+
+    this._currTabIndex = ev.detail.index;
+    if (!this._historyImported) {
+      import("./ha-more-info-tab-history");
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts:255]: Let's import this in `firstUpdated`. That will make sure the dialog is loaded and rendered fast, and that if you go to the history tab it is also there immediately.

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      now.toISOString(),
+      this.entityId,
+      true
+    );
+
+    // @ts-ignore
+    this._entries = entries;
+  }
+
+  private async _fetchUserNames() {
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:128]: We got to think of something for this, not funny to load the same data on every more info window.

File: src/panels/logbook/ha-logbook.ts [hunk 8]
```diff
@@ -98,46 +101,49 @@ class HaLogbook extends LitElement {
...
+            <ha-icon
+              .icon=${state ? stateIcon(state) : domainIcon(item.domain)}
+            ></ha-icon>
+            <div class="message">
+              ${!item.entity_id
+                ? html` <span class="name">${item.name}</span> `
+                : html`
+                    <a
+                      href="#"
+                      @click=${this._entityClicked}
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:114]: We don't want to open the more info window from the more info window?

File: src/dialogs/more-info/ha-more-info-tab-history.ts (hunk 7 above)
Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts:128]: Yea. Do you have an idea?

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 9]
```diff
@@ -130,38 +128,18 @@ export class MoreInfoTabHistoryDialog extends LitElement {
...
-    }
-    for (const person of persons.config) {
-      if (person.user_id) {
-        userIdToName[person.user_id] = person.name;
-      }
-    }
+  private _fetchPersonNames() {
+    const personEntities = Object.keys(this.hass.states).filter((entityId) =>
+      entityId.startsWith("person")
+    );
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:134]:
```suggestion
    Object.values(this.hass.states).forEach((entity) =>
      if (computeStateDomain(entity) === "person" && entity.attributes.user_id) {
        this._persons[person.attributes.user_id] = person.attributes.friendly_name;
      }
    );
```

**Comment by bdraco** [comment]
Screen_Shot_2020-09-02_at_1_32_01_PM

Maybe suppress the name since we are already in a modal that is only about the name?

**Comment by zsarnett** [comment]
image

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 10]
```diff
@@ -256,8 +268,6 @@ export class MoreInfoDialog extends LitElement {
           --mdc-theme-on-primary: var(--primary-text-color);
           --mdc-theme-primary: var(--mdc-theme-surface);
           flex-shrink: 0;
-          border-bottom: 1px solid
-            var(--mdc-dialog-scroll-divider-color, rgba(0, 0, 0, 0.12));
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts:260]: We should still show this border when we don't show the tabs

File: src/dialogs/more-info/ha-more-info-dialog.ts [hunk 11]
```diff
@@ -268,6 +278,15 @@ export class MoreInfoDialog extends LitElement {
...
         }
 
+        mwc-tab-bar {
+          border-bottom: 1px solid
+            var(--mdc-dialog-scroll-divider-color, rgba(0, 0, 0, 0.12));
+        }
+
+        .progress-wrapper {
+          height: 50px;
+        }
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-dialog.ts:288]: unused?

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+    const now = new Date();
+    const entries = await getLogbookData(
+      this.hass,
+      yesterday.toISOString(),
+      now.toISOString(),
+      this.entityId,
+      true
+    );
+
+    // @ts-ignore
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:129]: Why?

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+      },
+      this.hass!.localize,
+      this.hass!.language
+    );
+  }
+
+  private async _getLogBookData() {
+    const yesterday = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
+    const now = new Date();
+    const entries = await getLogbookData(
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:121]:
```suggestion
    this._entries = await getLogbookData(
```

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    }
+
+    if (changedProps.has("entityId")) {
+      this._refreshData();
+    }
+  }
+
+  private async _refreshData(): Promise<void> {
+    this._isLoading = true;
+    this._fetchPersonNames();
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:90]: We only need to do this when person entities changed

File: src/panels/logbook/ha-logbook.ts [hunk 15]
```diff
@@ -193,10 +223,20 @@ class HaLogbook extends LitElement {
...
+
       a {
         color: var(--primary-color);
       }
 
+      :host([no-click]) a {
+        color: inherit;
+        text-decoration: none;
+        cursor: auto;
+      }
```
Comment (bramkragten) [inline src/panels/logbook/ha-logbook.ts:238]: Oh that was done here :-)

File: src/dialogs/more-info/ha-more-info-tab-history.ts (hunk 14 above)
Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts:90]: How do we know they changed?

Comment (bramkragten) [reply src/dialogs/more-info/ha-more-info-tab-history.ts:90]: Check for state changes, but might not be worth the trouble, never mind.

Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts]: Should I just fetch it once on first updated and not worry about it?

File: src/dialogs/more-info/ha-more-info-tab-history.ts [hunk 16]
```diff
@@ -27,78 +27,64 @@ export class MoreInfoTabHistoryDialog extends LitElement {
...
+      <ha-logbook
+        narrow
+        no-icon
+        no-name
+        class=${classMap({ "has-entries": Boolean(this._entries?.length) })}
+        .hass=${this.hass}
+        .entries=${this._entries || []}
+        .userIdToName=${this._persons}
+        .loading=${!this._entries}
+      ></ha-logbook>
```
Comment (bramkragten) [inline src/dialogs/more-info/ha-more-info-tab-history.ts:62]: Wouldn't this be better?
```suggestion
      ${!this._entries ? html`<spinner></spinner>` : this._entries.length ? html`<ha-logbook
        narrow
        no-icon
        no-name
        .hass=${this.hass}
        .entries=${this._entries}
        .userIdToName=${this._persons}
        .loading=${!this._entries}
      ></ha-logbook>` : ""}
```

Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts:62]: Well the log book covers length of 0 entries

Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts:62]: I just mimicked the state history charts
but you are probably right that its better in the parent element

Comment (zsarnett) [reply src/dialogs/more-info/ha-more-info-tab-history.ts]: I reset and forced pushed the changes

**Review by bramkragten — APPROVED** [review]

**Pull request merged**